#713 improve quote handling

jseter opened the pull request:
Fix substring being called with invalid indexes.  Added strict quote mode to treat malformed quote as unquoted field.  remove depreciated substr usage.

pokoli commented:
Hi, thanks for your proposal. 

Could you explain which is the objective of this PR and if which issue is fixing (if any).

jseter commented:
The primary objective is to address #713, where a malformed quoted input gets into a long loop that freezes the browser.  I added a strict quoting mode to fail as early as possible and treat the malformed input as an unquoted field.

Additionally while examining the code I found in cases where a quoted string contains the newline or delimiter, the current value of those indexes are stale when used by `extraSpaces()` which depends on the substring between the closing quote and the end of the field or record.  The default behavior of substring in this condition is to just flip its inputs which does not produce the desired result.
I added test cases to catch this problem.

Adjusted the index reported on quoting errors to represent the start of the quote, instead of the first character inside the quote.

limited the "deepEqual" undefined issue to address only the serialization of the result, instead of depending on JSON stringify obeying a specific ordering as the string compare would require.

Refactored to not use substr which is a legacy function and should be treated as depreciated.

jseter commented:
My test file still manages to freeze the browser, so more investigation is probably needed to handle this particular issue.

jseter commented:
browser is using minified file which isn't updated in my test.  using the non-minified file so far is working as expected.

pokoli commented:
> My test file still manages to freeze the browser, so more investigation is probably needed to handle this particular issue.

You should use `npm test` to corretcly run the tests using the proper version.

jseter commented:
yes, i used `npm test` during the development.  I did not submit until all tests passed.  What I meant is that I tried to use the library in an actual application against the file that raised the issue, but the minified file is not automatically generated during pack or test so the browser was using the old unchanged code.

pokoli commented:
> yes, i used `npm test` during the development. I did not submit until all tests passed. What I meant is that I tried to use the library in an actual application against the file that raised the issue, but the minified file is not automatically generated during pack or test so the browser was using the old unchanged code.

The minified file always includes the latest relase code and it's updated on the release process.

pokoli commented on papaparse.js:
```diff
@@ -1092,7 +1099,7 @@ License: MIT
 			_paused = true;
 			_parser.abort();
-			_input = _input.substr(_parser.getCharIndex());
+			_input = _input.substring(_parser.getCharIndex());
```
Why changing this?

jseter replied:
Refactored to not use substr which is a legacy function and should be treated as depreciated.

pokoli replied:
Ok, could you please move this to a separate PR which does not modify the test suite?

pokoli commented on papaparse.js:
```diff
@@ -1330,7 +1337,7 @@ License: MIT
 		function guessLineEndings(input, quoteChar)
 		{
-			input = input.substr(0, 1024 * 1024);	// max length 1 MB
+			input = input.substring(0, 1024 * 1024);	// max length 1 MB
```
Idem, why changing?

pokoli commented on papaparse.js:
```diff
@@ -1382,6 +1389,8 @@ License: MIT
 		var preview = config.preview;
 		var fastMode = config.fastMode;
+		var temp1;
+		var strictQuote = (temp1 = config.strictQuote) !== undefined ? temp1 : false;
```
This line is a litte bit strange and hard to understand. Could you explain why it is needed?

jseter replied:
instead of fetching strictQuote from config once to test it for being undefined and again to use the value (which could have changed between the two points if the object installed a getter method, it saves it to a temporary local.
```
var strictQuote;

var temp1 = config.strictQuote;
if(temp1 !== undefined) {
    strictQuote = temp1;
} else {
    strictQuote = false; // default value
}
```

pokoli commented on papaparse.js:
```diff
@@ -1731,7 +1774,7 @@ License: MIT
 
 			/** Gets the delimiter character, which is not inside the quoted field */
-			function getNextUnqotedDelimiter(nextDelim, quoteSearch, newLine) {
+			function getNextUnquotedDelimiter(nextDelim, quoteSearch, newLine) {
```
Mmm, I will be happy to merge this as a separate PR as it does not seem releated.

pokoli commented on tests/test-cases.js:
```diff
@@ -22,6 +22,61 @@ try {
 	XHR_ENABLED = true;
 } catch (e) {} // safari, ie
 
+var __substring = String.prototype.substring;
```
Why is this all stuff neeeded?

jseter replied:
it allows the unit test to fail if substring is called with bad inputs, which is the first issue being addressed.

pokoli commented on tests/test-cases.js:
```diff
@@ -1475,7 +1580,7 @@ describe('Parse Tests', function() {
 				assert.deepEqual(actual.meta, test.expected.meta);
 			}
-			assert.deepEqual(JSON.stringify(actual.errors), JSON.stringify(test.expected.errors));
+			assert.deepEqual(JSON.parse(JSON.stringify(actual.errors)), test.expected.errors);
```
Why changing this?

jseter replied:
as covered in the explanation, JSON does not guarantee ordering or the properties within an object so doing a string compare of two JSON strings is not a good practice.  The goal of the JSON hack is to eliminate the `undefined` properties which deepEqual strictly would fail this on.  I have addressed in a separate branch making PapaParse not produce the undefined value so that this wouldn't be needed in that branch.
This modification keeps the value assertion `test.expected.errors` left untampered while using the JSON workaround for its intended purpose and avoiding the ordering problem.

pokoli commented on tests/test-cases.js:
```diff
@@ -246,14 +317,14 @@ var CORE_PARSER_TESTS = [
 				"message": "Trailing quote on quoted field is malformed",
 				"row": 0,
-				"index": 3
+				"index": 2
```
Why changing the index?

jseter replied:
as mentioned in the explanation, the index should represent the beginning of the quote or column and not inside the quoted string.

jseter commented:
Will break into separate PR's to simplify.
